arcjet-bun/index.ts

```diff
@@ -249,12 +153,108 @@ export default function arcjet<const Rules extends (Primitive | Product)[]>(
 ): ArcjetBun<Simplify<ExtraProps<Rules>>> {
   const client = options.client ?? createRemoteClient();
 
+  // Assuming the `handler()` function was used around Bun's fetch handler this
+  // WeakMap should be populated with IP addresses inspected via
+  // `server.requestIP()`
+  const ipCache = new WeakMap<Request, string>();
+
   const log = options.log
     ? options.log
     : new Logger({
         level: logLevel(env),
       });
 
+  function toArcjetRequest<Props extends PlainObject>(
+    request: Request,
+    props: Props,
+  ): ArcjetRequest<Props> {
+    const cookies = request.headers.get("cookie") ?? undefined;
+
+    // We construct an ArcjetHeaders to normalize over Headers
+    const headers = new ArcjetHeaders(request.headers);
+
+    const url = new URL(request.url);
+    let ip = findIP(
+      {
+        // This attempts to lookup the IP in the `ipCache`. This is primarily a
+        // workaround to the API design in Bun that requires access to the
+        // `Server` to lookup an IP.
+        ip: ipCache.get(request),
+      },
+      headers,
+      { platform: platform(env) },
+    );
+    if (ip === "") {
+      // If the `ip` is empty but we're in development mode, we default the IP
+      // so the request doesn't fail.
+      if (isDevelopment(env)) {
+        log.warn("Using 127.0.0.1 as IP address in development mode");
+        ip = "127.0.0.1";
+      } else {
+        log.warn(
+          "Empty IP address in non-development mode. Did you mean to set ARCJET_ENV=development?",
```

Comment:
Reworded slightly to make it clearer:

```suggestion
          "Client IP address is missing. If this is a dev environment set ARCJET_ENV=development in your .env.local file.",
```

I'm also wondering if we should also point people to [the docs](https://docs.arcjet.com/troubleshooting#generatefingerprint-ip-is-empty--failed_precondition-client-ip-not-provided) where we have a more detailed expanation. That URL is quite long so we could provide shorter URL redirects for error messages. Next.js does this quite effectively e.g. https://nextjs.org/docs/messages/prerender-error

Reply:
Since not everyone will use `.env` files (I'm thinking primarily of dockerfiles here), I don't think we should explicitly mention that file. How about:

```suggestion
          `Client IP address is missing. If this is a dev environment set the ARCJET_ENV env var to "development"`,
```

Reply:
Sounds good - I'm happy with your suggested change.